File: resources/[core]/dg-core/client/blipmanager.lua
DGCore.Blips = {};
DGCore.Blips.disabledCategory = {}
DGCore.Blips.blipStore = {}
DGCore.Blips.infoStore = {}

local isCatDisabled = function(cat)
  for k,_ in pairs(DGCore.Blips.disabledCategory) do
    if k == cat then
      return true
    end
  end
  return false
end

local createBlip = function(name, cat, info)
  local newBlip = AddBlipForCoord(info.coords.x, info.coords.y, info.coords.z)
  SetBlipSprite(newBlip, info.sprite)
  SetBlipScale(newBlip, info.scale)
  SetBlipAsShortRange(newBlip, info.isShortRange)
  -- Default to white
  SetBlipColour(newBlip, info.color)
  SetBlipDisplay(newBlip, info.display)

	BeginTextCommandSetBlipName("STRING")
	AddTextComponentString(info.text)
	EndTextCommandSetBlipName(newBlip)

  DGCore.Blips.blipStore[cat][name] = newBlip
end

local removeBlip = function(id, cat)
  if not DGCore.Blips.blipStore[cat] or not DGCore.Blips.blipStore[cat][id] then
    return
  end
  RemoveBlip(DGCore.Blips.blipStore[cat][id])
end

DGCore.Blips.Add = function(category, data)
  if (not DGCore.Blips.infoStore[category]) then
    DGCore.Blips.infoStore[category] = {}
  end
  if (not DGCore.Blips.blipStore[category]) then
    DGCore.Blips.blipStore[category] = {}
  end

  -- Remove blip if one already exists with id
  removeBlip(data.id, category)

  DGCore.Blips.infoStore[category][data.id] = {
    category = category,
    text = data.text,
    coords = data.coords,
    sprite = data.sprite,
    isShortRange = (data.range ~= nil and data.range) or true,
    scale = data.scale or 0.8,
    color = data.color or 4,
    display = data.display or 6,
  }

  if isCatDisabled(category) then
    return
  end
  createBlip(data.id, category, DGCore.Blips.infoStore[category][data.id])
end

DGCore.Blips.enableCategory = function(cat)
  if not isCatDisabled() then
    return
  end
  DGCore.Blips.disabledCategory[cat] = nil
  if not DGCore.Blips.infoStore[cat] then
    return
  end
  for id, info in pairs(DGCore.Blips.infoStore[cat]) do
    createBlip(id, cat, info)
  end
end

DGCore.Blips.disableCategory = function(cat)
  if isCatDisabled() then
    return
  end
  DGCore.Blips.disabledCategory[cat] = true
  if not DGCore.Blips.infoStore[cat] then
    return
  end
  for id, _ in pairs(DGCore.Blips.infoStore[cat]) do
    removeBlip(id, cat)
  end
end

DGCore.Blips.removeCategory = function(cat)
  for id, _ in pairs(DGCore.Blips.infoStore[cat]) do
    removeBlip(id, cat)
  end
  DGCore.Blips.infoStore[cat] = nil
  DGCore.Blips.blipStore[cat] = nil
end

File: resources/[dg]/dg-misc/typescript/src/client/services/blipmanager.ts
const disabledCategory = new Set<string>();
const blips: Record<string, NBlipManager.Info & { category: string; blip: number | null }> = {};

const createBlip = (id: string, info: NBlipManager.Info) => {
  let newBlip: number;
  if ('radius' in info) {
    newBlip = AddBlipForRadius(info.coords.x, info.coords.y, info.coords.z, info.radius);
  } else {
    newBlip = AddBlipForCoord(info.coords.x, info.coords.y, info.coords.z);
    SetBlipSprite(newBlip, info.sprite);
    SetBlipScale(newBlip, info.scale ?? 1);
    SetBlipAsShortRange(newBlip, info.isShortRange ?? true);

    if (info.text) {
      BeginTextCommandSetBlipName('STRING');
      AddTextComponentString(info.text);
      EndTextCommandSetBlipName(newBlip);
    }
  }

  SetBlipColour(newBlip, info.color ?? 0);
  SetBlipDisplay(newBlip, info.display ?? 6);

  blips[id].blip = newBlip;
};

const destroyBlip = (id: string) => {
  const info = blips[id];
  if (!info || !info.blip) return;
  RemoveBlip(info.blip);
  info.blip = null;
};

global.exports('addBlip', (data: NBlipManager.CreateData) => {
  const { id, ...info } = data;

  // Destroy if already exists
  destroyBlip(id);

  blips[id] = { ...info, blip: null };

  if (!disabledCategory.has(info.category)) {
    createBlip(id, info);
  }
});

global.exports('removeBlip', (id: string) => {
  destroyBlip(id);
  delete blips[id];
});

global.exports('enableCategory', (category: string) => {
  const wasDisabled = disabledCategory.delete(category);
  if (!wasDisabled) return;

  for (const [id, info] of Object.entries(blips)) {
    if (info.category !== category) continue;
    if (info.blip) continue;
    createBlip(id, info);
  }
});

global.exports('disableCategory', (category: string) => {
  if (disabledCategory.has(category)) return;
  disabledCategory.add(category);

  for (const [id, info] of Object.entries(blips)) {
    if (info.category !== category) continue;
    if (!info.blip) continue;
    destroyBlip(id);
  }
});

global.exports('removeCategory', (category: string) => {
  // cache or we will have issues by modifying blips obj while iterating
  const idsToRemove: string[] = [];

  for (const [id, info] of Object.entries(blips)) {
    if (info.category !== category) continue;
    idsToRemove.push(id);
  }

  for (const id of idsToRemove) {
    destroyBlip(id);
    delete blips[id];
  }
});
